packages/web3-eth-abi/src/types.ts

```diff
@@ -1,57 +1,207 @@
-	type: 'constructor';
-	stateMutability: 'nonpayable' | 'payable';
-}
+export type AbiBaseFragment = {
+	readonly type: FragmentTypes;
+};
+
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiConstructorFragment = AbiBaseFragment & {
+	readonly type: 'constructor';
+	readonly stateMutability: 'nonpayable' | 'payable';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+};
+
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiFunctionFragment = AbiBaseFragment & {
+	readonly name: string;
+	readonly type: 'function';
+	readonly stateMutability: 'nonpayable' | 'payable' | 'pure' | 'view';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+	readonly outputs: ReadonlyArray<AbiParameter>;
+
+	readonly constant?: boolean; // stateMutability == 'pure' or stateMutability == 'view'
+	readonly payable?: boolean; // stateMutability == 'payable'
+};
 
-export interface JsonAbiFunctionFragment extends JsonAbiBaseFragment {
-	type: 'function';
-	stateMutability: 'nonpayable' | 'payable' | 'pure' | 'view';
-	outputs?: Array<JsonAbiParameter>;
+export type AbiFallbackFragment = AbiBaseFragment & {
+	readonly name?: never;
+	readonly type: 'fallback';
+	readonly stateMutability: 'nonpayable' | 'payable' | 'pure' | 'view';
+	readonly inputs?: never;
+	readonly outputs?: never;
 
 	// legacy properties
-	constant?: boolean; // stateMutability == 'pure' or stateMutability == 'view'
-	payable?: boolean; // stateMutability == 'payable'
-}
+	readonly constant?: boolean; // stateMutability == 'pure' or stateMutability == 'view'
+	readonly payable?: boolean; // stateMutability == 'payable'
+};
 
-export interface JsonAbiEventFragment extends JsonAbiBaseFragment {
-	type: 'event';
-	anonymous?: boolean;
-}
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiEventFragment = AbiBaseFragment & {
+	readonly name: string;
+	readonly type: 'event';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+	readonly anonymous?: boolean;
+};
 
 // https://docs.soliditylang.org/en/latest/abi-spec.html#json
-export type JsonAbiFragment =
-	| JsonAbiConstructorFragment
-	| JsonAbiFunctionFragment
-	| JsonAbiEventFragment;
+export type AbiFragment =
+	| AbiConstructorFragment
+	| AbiFunctionFragment
+	| AbiEventFragment
+	| AbiFallbackFragment;
+
+export type ContractAbi = ReadonlyArray<AbiFragment>;
+
+export type AbiInput = string | AbiParameter | { readonly [key: string]: unknown };
+
+export type FilterAbis<Abis extends ContractAbi, Filter, Abi = Abis[number]> = Abi extends Filter
+	? Abi
+	: never;
+
+type _TypedArray<Type, Size extends string> = Size extends ''
+	? Type[]
+	: FixedSizeArray<Type, ConvertToNumber<Size>>;
+
+export type PrimitiveAddressType<Type extends string> = Type extends `address[${infer Size}]`
+	? _TypedArray<Address, Size>
+	: Type extends 'address'
+	? Address
+	: never;
+
+export type PrimitiveStringType<Type extends string> = Type extends `string${string}[${infer Size}]`
```

Comment:
I would love to have extensive JSDocs here explaining the rationale of this types. For external contributors this is hard to read and comments would help a lot

Reply:
Are there benchmarks of the compilation cost of extensive use of Template Literal Types?

Reply:
Never found any such, but these types provides real-time developer feedback in IDE so fast enough to use.

Reply:
Is there any demo project I can try with this feature?